AST: Add support for `public` and `@usableFromInline` access notes

tshortli opened the pull request:
Resolves rdar://103411848

tshortli commented:
@swift-ci please test

beccadax commented on test/SILGen/objc_access_notes.swift:
```diff
@@ -551,6 +551,15 @@ func useWotsit(_ w: Wotsit<String>) {
 
 func other() { }
 
+// expected-remark@+2 {{implicitly added '@usableFromInlin' to this struct, as specified by access note for fancy test suite}}
+// expected-note@+1 {{add '@usableFromInlin' explicitly to silence this warning}} {{1-1=@usableFromInline}}
```
The attribute name in these diagnostics is misspelled—if they pass, something must be wrong elsewhere.

tshortli replied:
Yup, there is indeed something wrong. I'll investigate

tshortli replied:
Fixed. The existing logic for pretty printing the attributes for diagnostics assumed too much about the output of `ASTPrinter`.

beccadax reviewed (commented):
There’s an important difference between the `Public` and `UsableFromInline` access note attributes and the others we’ve implemented before. If you use `ObjC: true` on a declaration that can’t be made `@objc`, the diagnostic for that is reduced to a warning; that won’t be the case for `Public: true` or `UsableFromInline: true`, so a bad access note could block the build of the project it’s applied to. But if we’re okay with that, then this PR looks fine (other than the diagnostic issue I mention below).

tshortli committed:
Sema: Fix printing of attributes in access note diagnostics.

Trim spaces from the result of prettyPrintAttrs() rather than unconditionally dropping the last character.

tshortli commented:
@swift-ci please smoke test and merge

slavapestov commented:
Unfortunately all this new access note logic is wrong. You're adding and removing attributes in TypeCheckDeclPrimary.cpp, so if you have a declaration in a secondary file with an access note applied to it, and you reference it from a primary file, the primary frontend job won't see the new attributes because they will never be applied.

You need to actually change AccessLevelRequest, IsObjCRequest and other requests to query access notes as needed. Don't mutate the AST in TypeCheckDeclPrimary; the purpose of that pass is to evaluate requests which emit diagnostics.

slavapestov commented on lib/AST/AccessRequests.cpp:
```diff
@@ -40,6 +41,9 @@ AccessLevel
   assert(!D->hasAccess());
 
+  // Access notes may add attributes that affect this calculus.
+  (void)evaluateOrDefault(evaluator, ApplyAccessNoteRequest{D}, {});
```
ApplyAccessNoteRequest also introduces DynamicAttr, UsableFromInlineAttr and ObjCAttr so doing it here doesn't make any sense as there's no guarantee that someone won't query those attributes before checking the access level.

slavapestov replied:
Specifically, what you need to do is the following:
- change the places that directly check for UsableFromInlineAttr to call a method on ValueDecl which evaluates a new UsableFromInlineRequest
- split up ApplyAccessNoteRequest
- in AccessLevelRequest, apply the access level from the API note
- in DynamicRequest, apply the dynamic attribute from the API note
- in ObjCRequest, apply the objc attribute from the API note
- in your new UsableFromInlineRequest, apply the usableFromInline attribute from the API note

Commit:
Sema: Allow accessnote to add `@usableFromInline` and `public` keywords.

Resolves rdar://103411848
